docs/security-best-practices.md

```diff
@@ -153,6 +153,87 @@ To change the default endpoint to be `localhost`-bound in all components, enable
 
 If `localhost` resolves to a different IP due to your DNS settings then explicitly use the loopback IP instead: `127.0.0.1` for IPv4 or `::1` for IPv6. In IPv6 setups, ensure your system supports both IPv4 and IPv6 loopback addresses to avoid issues.
 
+Using `localhost` may not work in environments like Docker, Kubernetes, and other environments that have non-standard networking setups.  We've documented working example setups for some of these environments:
+
+
+#### Docker
```

Comment:
This seems to be about Docker Compose specifically
```suggestion
#### Docker Compose
```

Reply:
Sort of	- but not quite.  I use the docker compose file format to demonstrate how to start the docker container but I could replace that with at `docker run` invocation and it should be almost identical behavior-wise.  The points I'm trying to make is that 1) if you run the collector inside docker, you can still connect to it by binding the port to a host port, and that 2) If you run the collector container and another container trying to reach the collector on the same docker network you can use the service name to find it.

Reply:
I can't make it work with `docker run`:

<details>
<summary>Collector configuration</summary>

```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: container-name:4317
      http:
        endpoint: container-name:4318

exporters:
  debug:
    verbosity: detailed

service:
  pipelines:
    logs:
      receivers: [otlp]
      exporters: [debug]
```

</details>

<details>
<summary>Logs</summary>

```
❯ sudo docker run -u 0 --name container-name -p 4317:4317 -p 4318:4318 -v $(pwd):/etc/otelcol otel/opentelemetry-collector:0.104.0 --config /etc/otelcol/collector.yaml
2024-07-11T15:20:17.839Z	info	service@v0.104.0/service.go:115	Setting up own telemetry...
2024-07-11T15:20:17.839Z	info	service@v0.104.0/telemetry.go:96	Serving metrics	{"address": ":8888", "level": "Normal"}
2024-07-11T15:20:17.840Z	info	exporter@v0.104.0/exporter.go:280	Development component. May change in the future.	{"kind": "exporter", "data_type": "logs", "name": "debug"}
2024-07-11T15:20:17.843Z	info	service@v0.104.0/service.go:193	Starting otelcol...	{"Version": "0.104.0", "NumCPU": 20}
2024-07-11T15:20:17.844Z	info	extensions/extensions.go:34	Starting extensions...
2024-07-11T15:20:17.844Z	info	otlpreceiver@v0.104.0/otlp.go:102	Starting GRPC server	{"kind": "receiver", "name": "otlp", "data_type": "logs", "endpoint": "container-name:4317"}
2024-07-11T15:20:17.856Z	info	service@v0.104.0/service.go:256	Starting shutdown...
2024-07-11T15:20:17.857Z	info	extensions/extensions.go:59	Stopping extensions...
2024-07-11T15:20:17.857Z	info	service@v0.104.0/service.go:270	Shutdown complete.
Error: cannot start pipelines: listen tcp: lookup container-name on 100.90.1.1:53: no such host
2024/07/11 15:20:17 collector server run finished with error: cannot start pipelines: listen tcp: lookup container-name on 100.90.1.1:53: no such host
```

</details>

Am I doing something wrong? I think it may be beneficial to have an explicit separate example for `docker run` (at least it's not clear to me what the equivalence is)

Reply:
Hmmmm you are right because docker compose creates a default network and then puts all the containers onto it... while docker run does not do that.  i will try and figure out what a `docker run` equivalence is